File: cicero_hours/loader.py
"""Read the CICERO 'Timer budsjettert og registrert pr. medarbeider' export.

The export is not a single CSV. It concatenates several tables, each preceded
by its own quoted header row, and the row offsets move between exports. This
module finds the header rows by signature instead of by position, so a new
export drops straight in.
"""

from __future__ import annotations

import io
from dataclasses import dataclass
from pathlib import Path

import pandas as pd

# A table is recognised by columns it must contain. First match wins, so put
# the more specific signatures first.
TABLE_SIGNATURES: list[tuple[str, set[str]]] = [
    ("registered", {"Medarbeider", "Prosjekt", "Hours - Reg.", "Year str"}),
    ("budget", {"Medarbeider", "Prosjekt", "Quantity - Hours", "Budget Type"}),
]


@dataclass
class RawExport:
    """The parsed blocks of one export file."""

    tables: dict[str, pd.DataFrame]
    unrecognised: list[pd.DataFrame]
    source: Path

    def require(self, name: str) -> pd.DataFrame:
        if name not in self.tables:
            found = ", ".join(sorted(self.tables)) or "none"
            raise KeyError(
                f"No '{name}' table in {self.source.name}. Recognised tables: {found}. "
                "If the export format changed, update TABLE_SIGNATURES in loader.py."
            )
        return self.tables[name]


def _split_blocks(text: str) -> list[list[str]]:
    """Split the file into blocks, each starting at a line that looks like a header.

    A header line is taken to be one whose every field is a bare quoted string
    and which is followed by at least one data line. That is deliberately loose:
    the point is to survive column reordering and added columns.
    """
    lines = text.splitlines()
    starts: list[int] = []
    for i, line in enumerate(lines):
        if not line.strip():
            continue
        fields = _sniff_fields(line)
        if fields is None:
            continue
        # Header rows in this export are all-text and contain no pure numbers.
        if any(_looks_numeric(f) for f in fields):
            continue
        if len(fields) >= 2:
            starts.append(i)
    if not starts:
        return []
    # Consecutive candidate lines are data, not a run of headers: keep the first
    # of each run only if the run length is 1.
    kept = [s for j, s in enumerate(starts) if j == 0 or s != starts[j - 1] + 1]
    kept = [s for s in kept if s + 1 not in starts or True]
    blocks: list[list[str]] = []
    for j, s in enumerate(kept):
        end = kept[j + 1] if j + 1 < len(kept) else len(lines)
        block = [ln for ln in lines[s:end] if ln.strip()]
        if len(block) > 1:
            blocks.append(block)
    return blocks


def _sniff_fields(line: str) -> list[str] | None:
    try:
        parsed = pd.read_csv(io.StringIO(line), header=None, dtype=str)
        row = next(iter(parsed.itertuples(index=False)))
    except Exception:
        return None
    return ["" if pd.isna(v) else str(v) for v in row]


def _looks_numeric(value: str) -> bool:
    v = value.strip().replace(" ", "").replace(",", ".")
    if not v:
        return False
    try:
        float(v)
    except ValueError:
        return False
    return True


def load_export(path: str | Path) -> RawExport:
    """Parse an export file into named tables."""
    path = Path(path)
    text = path.read_text(encoding="utf-8-sig")
    tables: dict[str, pd.DataFrame] = {}
    unrecognised: list[pd.DataFrame] = []

    for block in _split_blocks(text):
        try:
            df = pd.read_csv(io.StringIO("\n".join(block)))
        except Exception:
            continue
        cols = set(df.columns)
        for name, signature in TABLE_SIGNATURES:
            if signature <= cols and name not in tables:
                tables[name] = df
                break
        else:
            unrecognised.append(df)

    return RawExport(tables=tables, unrecognised=unrecognised, source=path)

File: web/src/loader.js
// Parse the CICERO 'Timer budsjettert og registrert pr. medarbeider' export.
//
// The export concatenates several tables into one file, each with its own
// quoted header row, and the row offsets move between exports. This module
// finds tables by column signature rather than position, so a new export
// drops straight in. Mirror of cicero_hours/loader.py, one for one, so a
// silent divergence at parse time cannot creep in.

import { TABLE_SIGNATURES } from './rules.js';

// {tables: {name: {columns, rows}}, unrecognised: [{columns, rows}]}
export function loadExport(text) {
  const tables = {};
  const unrecognised = [];
  for (const block of splitBlocks(stripBom(text))) {
    const parsed = parseBlock(block);
    if (!parsed) continue;
    const cols = new Set(parsed.columns);
    let matched = null;
    for (const sig of TABLE_SIGNATURES) {
      if (!(sig.name in tables) && isSubset(sig.columns, cols)) {
        matched = sig.name;
        break;
      }
    }
    if (matched) tables[matched] = parsed;
    else unrecognised.push(parsed);
  }
  return { tables, unrecognised };
}

export function requireTable(raw, name) {
  if (!(name in raw.tables)) {
    const found = Object.keys(raw.tables).sort().join(', ') || 'none';
    throw new Error(
      `No '${name}' table in the export. Recognised tables: ${found}. ` +
      `If the export format changed, update table_signatures in spec/rules.json.`
    );
  }
  return raw.tables[name];
}

function stripBom(text) {
  return text.charCodeAt(0) === 0xFEFF ? text.slice(1) : text;
}

function isSubset(sub, sup) {
  for (const x of sub) if (!sup.has(x)) return false;
  return true;
}

// Header lines are all-text and contain at least two fields; data lines carry
// at least one numeric-looking value. Loose on purpose, so column reorder or
// added columns still find the right block.
function splitBlocks(text) {
  const lines = text.split(/\r\n|\n|\r/);
  const starts = [];
  for (let i = 0; i < lines.length; i++) {
    if (!lines[i].trim()) continue;
    const fields = sniffFields(lines[i]);
    if (!fields) continue;
    if (fields.some(looksNumeric)) continue;
    if (fields.length >= 2) starts.push(i);
  }
  if (!starts.length) return [];
  // A run of consecutive candidate lines is data, not a stack of headers.
  const kept = starts.filter((s, j) => j === 0 || s !== starts[j - 1] + 1);
  const blocks = [];
  for (let j = 0; j < kept.length; j++) {
    const s = kept[j];
    const end = j + 1 < kept.length ? kept[j + 1] : lines.length;
    const block = lines.slice(s, end).filter(l => l.trim());
    if (block.length > 1) blocks.push(block);
  }
  return blocks;
}

function parseBlock(block) {
  const columns = parseCsvLine(block[0]);
  if (!columns.length) return null;
  const rows = block.slice(1).map(line => {
    const fields = parseCsvLine(line);
    const row = {};
    for (let i = 0; i < columns.length; i++) {
      row[columns[i]] = fields[i] === undefined ? '' : fields[i];
    }
    return row;
  });
  return { columns, rows };
}

function sniffFields(line) {
  try { return parseCsvLine(line); } catch { return null; }
}

// Norwegian numbers use a comma for the decimal, and spaces sometimes appear as
// thousands separators; normalise both before deciding "is this a number".
function looksNumeric(value) {
  const v = String(value).trim().replaceAll(' ', '').replaceAll(',', '.');
  if (!v) return false;
  const n = Number(v);
  return !Number.isNaN(n) && Number.isFinite(n);
}

// Minimal CSV line parser: double-quoted fields with embedded commas and
// doubled quotes ("") as an escape. The export has no fields with embedded
// newlines, and neither does the Python loader handle them.
function parseCsvLine(line) {
  const fields = [];
  let cur = '', inQuotes = false, i = 0;
  while (i < line.length) {
    const c = line[i];
    if (inQuotes) {
      if (c === '"') {
        if (line[i + 1] === '"') { cur += '"'; i += 2; continue; }
        inQuotes = false; i++;
      } else { cur += c; i++; }
    } else {
      if (c === '"') { inQuotes = true; i++; }
      else if (c === ',') { fields.push(cur); cur = ''; i++; }
      else { cur += c; i++; }
    }
  }
  fields.push(cur);
  return fields;
}
